rust/src/http2/parser.rs

```diff
@@ -0,0 +1,1483 @@
+            value: "".to_string(),
+        }),
+        44 => Some(HTTP2FrameHeaderBlock {
+            name: "last-modified".to_string(),
+            value: "".to_string(),
+        }),
+        45 => Some(HTTP2FrameHeaderBlock {
+            name: "link".to_string(),
+            value: "".to_string(),
+        }),
+        46 => Some(HTTP2FrameHeaderBlock {
+            name: "location".to_string(),
+            value: "".to_string(),
+        }),
+        47 => Some(HTTP2FrameHeaderBlock {
+            name: "max-forwards".to_string(),
+            value: "".to_string(),
+        }),
+        48 => Some(HTTP2FrameHeaderBlock {
+            name: "proxy-authenticate".to_string(),
+            value: "".to_string(),
+        }),
+        49 => Some(HTTP2FrameHeaderBlock {
+            name: "proxy-authorization".to_string(),
+            value: "".to_string(),
+        }),
+        50 => Some(HTTP2FrameHeaderBlock {
+            name: "range".to_string(),
+            value: "".to_string(),
+        }),
+        51 => Some(HTTP2FrameHeaderBlock {
+            name: "referer".to_string(),
+            value: "".to_string(),
+        }),
+        52 => Some(HTTP2FrameHeaderBlock {
+            name: "refresh".to_string(),
+            value: "".to_string(),
+        }),
+        53 => Some(HTTP2FrameHeaderBlock {
+            name: "retry-after".to_string(),
+            value: "".to_string(),
+        }),
+        54 => Some(HTTP2FrameHeaderBlock {
+            name: "server".to_string(),
+            value: "".to_string(),
+        }),
+        55 => Some(HTTP2FrameHeaderBlock {
+            name: "set-cookie".to_string(),
+            value: "".to_string(),
+        }),
+        56 => Some(HTTP2FrameHeaderBlock {
+            name: "strict-transport-security".to_string(),
+            value: "".to_string(),
+        }),
+        57 => Some(HTTP2FrameHeaderBlock {
+            name: "transfer-encoding".to_string(),
+            value: "".to_string(),
+        }),
+        58 => Some(HTTP2FrameHeaderBlock {
+            name: "user-agent".to_string(),
+            value: "".to_string(),
+        }),
+        59 => Some(HTTP2FrameHeaderBlock {
+            name: "vary".to_string(),
+            value: "".to_string(),
+        }),
+        60 => Some(HTTP2FrameHeaderBlock {
+            name: "via".to_string(),
+            value: "".to_string(),
+        }),
+        61 => Some(HTTP2FrameHeaderBlock {
+            name: "www-authenticate".to_string(),
+            value: "".to_string(),
+        }),
+        _ => None,
+    }
+}
+
+#[derive(Clone)]
+pub struct HTTP2FrameHeaderBlock {
+    pub name: String,
+    pub value: String,
+}
+
+fn http2_parse_headers_block_indexed(input: &[u8]) -> IResult<&[u8], HTTP2FrameHeaderBlock> {
+    //TODO ask why we need this over error[E0282]: type annotations needed for `((&[u8], usize), O1)`
+    fn parser(input: &[u8]) -> IResult<&[u8], (u8, HTTP2FrameHeaderBlock)> {
```

Comment:
Any rust/nom style advice here ?

Reply:
cc @jasonish

Reply:
If the question is about the style of function inside a function, I think not using nom macros, and instead using nom functions (new nom style) would give you a more pleasing style.  I have yet to master converting from macros to functions though.

Reply:
I had to use this function (based on some other parser in Suricata) because if I use instead of 

`let (i2, indexed) = parser(input)?;` 

the corresponding function

`let (i2, indexed) = bits!(...)?;`

It does not compile with error E0282

Reply:
Probably needs some type annotation like seen in this example:
https://docs.rs/nom/5.1.1/nom/bits/fn.bits.html#example

My quick attempts to convert it failed as well.

Reply:
Leaving it this way is fine I suppose then ?